lib/collection/src/shards/transfer/shard_transfer.rs

```diff
@@ -773,67 +760,72 @@ where
+                if attempt > 0 {
+                    sleep(RETRY_DELAY * attempt as u32).await;
+
+                    log::warn!(
+                        "Retrying shard transfer {} -> {} (retry {attempt})",
+                        transfer.shard_id,
+                        transfer.to,
+                    );
+                }
+
+                transfer_shard(
+                    transfer.clone(),
+                    shards_holder.clone(),
+                    consensus.as_ref(),
+                    collection_id.clone(),
+                    &collection_name,
+                    channel_service.clone(),
+                    &snapshots_path,
+                    &temp_dir,
+                )
+                .await
+            };
+
+            result = cancel::future::cancel_on_token(cancel.clone(), future)
+                .await
+                .map_err(Into::into)
+                .and_then(|res| res);
+
+            match &result {
+                Ok(()) => break,
+
+                // TODO: Should we break early if shard transfer is cancelled, or should we handle it as regular error?
+                Err(CollectionError::Cancelled { .. }) => break,
+
+                Err(err) => {
                     log::error!(
-                        "Failed to transfer shard {} -> {}: {error}",
+                        "Failed to transfer shard {} -> {}: {err}",
                         transfer.shard_id,
                         transfer.to,
                     );
 
-                    false
+                    // Revert queue proxy if we still have any to prepare for the next attempt
+                    if let Some(shard) = shards_holder.read().await.get_shard(&transfer.shard_id) {
+                        shard.revert_queue_proxy_local().await;
+                    }
                 }
-            };
-            if cancel.is_cancelled() {
-                return false;
-            }
-            if !finished {
-                tries -= 1;
-                log::warn!(
-                    "Retrying shard transfer {} -> {} (retry {})",
-                    transfer.shard_id,
-                    transfer.to,
-                    MAX_RETRY_COUNT - tries
-                );
-                let exp_timeout = RETRY_DELAY * (MAX_RETRY_COUNT - tries) as u32;
-                sleep(exp_timeout).await;
             }
         }
 
-        if finished {
-            // On the end of transfer, the new shard is active but most likely is under the optimization
-            // process. Requests to this node might be slow, but we rely on the assumption that
-            // there should be at least one other replica that is not under optimization.
-            on_finish.await;
-        } else {
-            on_error.await;
+        match &result {
+            Ok(()) => on_finish.await,
+
+            // TODO: Should we ignore `on_error` if shard transfer is cancelled, or should we handle it as regular error?
```

Comment:
I think this branch is fine; doing nothing on cancel.

Reply:
Removed.